internal/expr/scan.go

```diff
@@ -0,0 +1,196 @@
+func (re *ResultExpr) All() ([][]any, error) {
+	var rows [][]any
+
+	outputTypes := getTypes(re.outputs)
+
+	for {
+		if ok, err := re.Next(); err != nil {
+			return [][]any{}, err
+		} else if !ok {
+			break
+		}
+
+		outputStructs := make([]reflect.Value, len(outputTypes))
+
+		for i, t := range outputTypes {
+			// reflect.New returns a pointer to a struct. It is dereferenced with Elem().
+			outputStructs[i] = reflect.New(t).Elem()
+		}
+
+		err := re.decodeRow(outputStructs...)
+		if err != nil {
+			return [][]any{}, err
+		}
+
+		row := make([]any, len(outputTypes))
+		for i, s := range outputStructs {
+			row[i] = s.Interface()
+		}
+
+		rows = append(rows, row)
+	}
+
+	re.Close()
+	return rows, nil
+}
+
+// Next prepares the next result row for reading with the Scan method.
+func (re *ResultExpr) Next() (bool, error) {
+	ok := re.rows.Next()
+	if !ok && re.rows.Err() != nil {
+		return false, re.rows.Err()
+	}
+	return ok, nil
+}
+
+// Decode copies the columns in the current row into the fields of the structs in dests.
+// dests must contain all the structs mentioned in the query.
+func (re *ResultExpr) Decode(dests ...any) (err error) {
+	defer func() {
+		if err != nil {
+			err = fmt.Errorf("cannot decode expression: %s", err)
+		}
+	}()
+
+	destVals := []reflect.Value{}
+
+	for _, dest := range dests {
+		if dest == nil {
+			return fmt.Errorf("need valid struct, got nil")
+		}
+
+		destVal := reflect.ValueOf(dest)
+		if destVal.Kind() != reflect.Pointer {
+			return fmt.Errorf("need pointer to struct, got non-pointer of kind %s", destVal.Kind())
+		}
+
+		destVal = reflect.Indirect(destVal)
+
+		if destVal.Kind() != reflect.Struct {
+			return fmt.Errorf("need pointer to struct, got pointer to %s", destVal.Kind())
+		}
+
+		t := destVal.Type()
+
+		typeFound := false
+		for _, f := range re.outputs {
+			if f.structType == t {
+				typeFound = true
+				break
+			}
+		}
+
+		if !typeFound {
+			return fmt.Errorf("no output expression of type %s", t.Name())
+		}
+
+		destVals = append(destVals, destVal)
+	}
+	if err := re.decodeRow(destVals...); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// decodeRow copies a result row into the fields of the reflect.Values in dests.
+// All the structs mentioned in the query must be in dests.
+// All dests must be of kind reflect.Struct and must be addressable and settable.
+func (re *ResultExpr) decodeRow(dests ...reflect.Value) error {
+	cols, err := re.rows.Columns()
+	if err != nil {
+		return err
+	}
+
+	var typeDest = make(map[reflect.Type]reflect.Value)
+	for _, dest := range dests {
+		typeDest[dest.Type()] = dest
+	}
+
+	ptrs := []any{}
+
+	// 	sqlair columns are named as _sqlair_X, where X = 0, 1, 2,...
+	//  offset is the difference between X and the index i and where column _sqlair_X is located in cols.
+	//  It allows non-sqlair columns to be returned in the results.
+	offset := 0
+
+	for i, col := range cols {
+		if col == "_sqlair_"+strconv.Itoa(i-offset) {
```

Comment:
This seems more complex/expensive than expected. We were the ones putting the index in, which should match the actual index that we have internally. Why are we doing such manual string building comparisons instead of actually using the index that we've put in ourselves directly?